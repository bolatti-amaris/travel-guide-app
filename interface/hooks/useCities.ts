import { useEffect, useCallback } from 'react';
import { City } from '@/core/domain/entities/City';
import { useCitiesStore } from '../../core/interface/store/useCitiesStore';

export const useCities = () => {
  const {
    cities,
    isLoading,
    error,
    fetchCities,
    selectedCity,
    setSelectedCity,
  } = useCitiesStore();

  useEffect(() => {
    fetchCities();
  }, []);

  const selectCity = (city: City) => {
    setSelectedCity(city);
  };

  const reset = useCallback(() => {
    useCitiesStore.getState().reset();
  }, []);

  return {
    cities,
    isLoading,
    error,
    selectedCity,
    selectCity,
    reset,
  };
};
