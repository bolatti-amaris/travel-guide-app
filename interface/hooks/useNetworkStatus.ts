import NetInfo from '@react-native-community/netinfo';
import { useEffect, useState } from 'react';

export const useNetworkStatus = () => {
  const [isConnected, setIsConnected] = useState(true);

  useEffect(() => {
    const unsubscribe = NetInfo.addEventListener((state) => {
      setIsConnected(Boolean(state.isConnected));
    });

    return () => unsubscribe();
  }, []);

  return { isConnected };
};
