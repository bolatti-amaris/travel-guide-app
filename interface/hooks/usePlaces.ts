import { useCallback, useEffect } from 'react';
import { useCitiesStore } from '@/core/interface/store/useCitiesStore';
import { usePlacesStore } from '@/core/interface/store/usePlacesStore';

export const usePlaces = () => {
  const { selectedCity } = useCitiesStore();
  const { places, isLoading, error, fetchPlaces } = usePlacesStore();

  useEffect(() => {
    const shouldFetch =
      selectedCity &&
      selectedCity.key &&
      places.length === 0 &&
      !isLoading &&
      !error;

    if (shouldFetch) {
      fetchPlaces(selectedCity.key);
    }
  }, [selectedCity, places.length, isLoading, error, fetchPlaces]);

  const reset = useCallback(() => {
    usePlacesStore.getState().reset();
  }, []);

  return {
    places,
    isLoading,
    error,
    fetchPlaces,
    reset,
  };
};
