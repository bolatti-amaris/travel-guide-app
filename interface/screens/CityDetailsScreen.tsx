import React, { useEffect } from 'react';
import { View, Text, ScrollView } from 'react-native';
import { usePlaces } from '../hooks/usePlaces';
import { useCities } from '../hooks/useCities';
import Loader from '../shared/Loader';
import PlaceList from '../components/PlaceList';
import ErrorMessageProps from '../shared/ErrorMessage';


const CityDetailsScreen = () => {
  const { selectedCity } = useCities();
  const { places, isLoading, error, reset, fetchPlaces } = usePlaces();

useEffect(() => {
    return () => {
      reset();
    };
  }, []);

  if (!selectedCity) {
    return (
      <View className="flex-1 justify-center items-center bg-white">
        <Text className="text-red-600">City not found</Text>
      </View>
    );
  }

  return (
    <ScrollView className="flex-1 bg-white p-6">
      <Text className="text-3xl font-bold mb-2">{selectedCity.name}</Text>
      <Text className="text-lg text-gray-600 italic mb-2">{selectedCity.nativeName}</Text>
      <Text className="text-base text-gray-700 mb-1">Currency: {selectedCity.currency}</Text>
      <Text className="text-base text-gray-700 mb-4">Language: {selectedCity.language}</Text>

      {isLoading && <Loader message="Loading places..." />}

      {error && (
        <ErrorMessageProps
        message={error}
        onRetry={() => fetchPlaces(selectedCity.key)}
      />
      )}

      {places.length > 0 && (
        <PlaceList places={places} />
      )}
    </ScrollView>
  );
};

export default CityDetailsScreen;