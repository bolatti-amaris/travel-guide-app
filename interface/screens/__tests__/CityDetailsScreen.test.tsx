import React from 'react';
import { render } from '@testing-library/react-native';
import CityDetailsScreen from '../CityDetailsScreen';

import { useCities } from '../../hooks/useCities';
import { usePlaces } from '../../hooks/usePlaces';

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    setOptions: jest.fn(),
  }),
}));

jest.mock('../../hooks/useCities', () => ({
  useCities: jest.fn(),
}));

jest.mock('../../hooks/usePlaces', () => ({
  usePlaces: jest.fn(),
}));

jest.mock('../../shared/Loader', () => {
  const { Text } = require('react-native');
  return () => <Text testID="loader">Loader</Text>;
});

jest.mock('../../shared/Retry', () => {
  const { Text } = require('react-native');
  return ({ message }: { message: string }) => (
    <Text testID="retry">{message}</Text>
  );
});

jest.mock('../../components/PlaceList', () => {
  const { Text } = require('react-native');
  return ({ places }: { places: any[] }) => (
    <Text testID="place-list">{places.length} places</Text>
  );
});

const mockCity = {
  id: 1,
  key: 'paris',
  name: 'Paris',
  nativeName: 'Paris',
  currency: 'EUR',
  language: 'fr',
};

describe('CityDetailsScreen', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('shows "City not found" if no selected city', () => {
    (useCities as jest.Mock).mockReturnValue({ selectedCity: null });
    (usePlaces as jest.Mock).mockReturnValue({
      places: [],
      isLoading: false,
      error: null,
      reset: jest.fn(),
      fetchPlaces: jest.fn(),
    });

    const { getByTestId } = render(<CityDetailsScreen />);
    expect(getByTestId('no-city')).toBeTruthy();
  });

  it('renders Loader when loading', () => {
    (useCities as jest.Mock).mockReturnValue({ selectedCity: mockCity });
    (usePlaces as jest.Mock).mockReturnValue({
      places: [],
      isLoading: true,
      error: null,
      reset: jest.fn(),
      fetchPlaces: jest.fn(),
    });

    const { getByTestId } = render(<CityDetailsScreen />);
    expect(getByTestId('loader')).toBeTruthy();
  });

  it('renders Retry on error', () => {
    (useCities as jest.Mock).mockReturnValue({ selectedCity: mockCity });
    (usePlaces as jest.Mock).mockReturnValue({
      places: [],
      isLoading: false,
      error: 'Error!',
      reset: jest.fn(),
      fetchPlaces: jest.fn(),
    });

    const { getByTestId } = render(<CityDetailsScreen />);
    expect(getByTestId('retry')).toBeTruthy();
  });

  it('renders Retry when empty state', () => {
    (useCities as jest.Mock).mockReturnValue({ selectedCity: mockCity });
    (usePlaces as jest.Mock).mockReturnValue({
      places: [],
      isLoading: false,
      error: null,
      reset: jest.fn(),
      fetchPlaces: jest.fn(),
    });

    const { getByTestId } = render(<CityDetailsScreen />);
    expect(getByTestId('retry')).toBeTruthy();
  });

  it('renders PlaceList if places exist', () => {
    (useCities as jest.Mock).mockReturnValue({ selectedCity: mockCity });
    (usePlaces as jest.Mock).mockReturnValue({
      places: [{ name: 'Tour Eiffel', type: 'monument', coordinates: [1, 2] }],
      isLoading: false,
      error: null,
      reset: jest.fn(),
      fetchPlaces: jest.fn(),
    });

    const { getByTestId } = render(<CityDetailsScreen />);
    expect(getByTestId('place-list')).toBeTruthy();
  });

  it('matches snapshot', () => {
    (useCities as jest.Mock).mockReturnValue({ selectedCity: mockCity });
    (usePlaces as jest.Mock).mockReturnValue({
      places: [],
      isLoading: false,
      error: null,
      reset: jest.fn(),
      fetchPlaces: jest.fn(),
    });

    const tree = render(<CityDetailsScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
