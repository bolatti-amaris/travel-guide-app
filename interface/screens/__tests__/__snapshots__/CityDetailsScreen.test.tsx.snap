// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityDetailsScreen matches snapshot 1`] = `
<RCTScrollView
  className="flex-1 px-4 bg-background dark:bg-background-dark"
  contentContainerStyle={
    {
      "flexGrow": 1,
    }
  }
  testID="city-details"
>
  <View>
    <Text
      className="text-2xl font-work-medium text-text dark:text-text-dark italic mb-1"
      testID="city-native-name"
    >
      (
      Paris
      )
    </Text>
    <Text
      className="text-xl text-text dark:text-text-dark font-work-light mb-1"
      testID="city-currency"
    >
      Currency: 
      EUR
    </Text>
    <Text
      className="text-xl text-text dark:text-text-dark font-work-light mb-4"
      testID="city-language"
    >
      Language: 
      fr
    </Text>
    <Text
      testID="retry"
    >
      No places available
    </Text>
  </View>
</RCTScrollView>
`;
